src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringConcat.cs

```diff
@@ -346,6 +453,210 @@ private BoundExpression RewriteStringConcatenationManyExprs(SyntaxNode syntax, I
             return BoundCall.Synthesized(syntax, receiverOpt: null, initialBindingReceiverIsSubjectToCloning: ThreeState.Unknown, method, array);
         }
 
+        private bool TryRewriteStringConcatenationWithSpanBasedConcat(SyntaxNode syntax, ImmutableArray<BoundExpression> args, [NotNullWhen(true)] out BoundExpression? result)
+        {
+            // As we scan arguments we might face span-based string.Concat. If it got to this stage it must be a user-provided one (since we weren't able to unwrap it back to strings previously)
+            // We try our best to merge it with another argument to minimize amount of intermediate string allocations,
+            // e.g. turn `string.Concat(span1, span2) + string3` into `string.Concat(span1, span2, string3ConvertedToSpan)`
+            // instead of `string.Concat(string.Concat(span1, span2), string3)`. We do that by separately tracking arguments as is
+            // and arguments if we unwrap all user-defined `string.Concat`s. If at the end amount of arguments with unwrapped user-defined `string.Concat`s
+            // is < 5 and we have respective span-based concat member, we emit it, otherwise treat user-provided `string.Concat`s as ordinary arguments
+            var preparedArgs = ArrayBuilder<BoundExpression>.GetInstance(capacity: args.Length);
+            var preparedArgsIfUnwrapUserStringConcat = ArrayBuilder<BoundExpression>.GetInstance(capacity: args.Length);
+
+            var needsSpanRefParamConstructor = false;
+            var needsImplicitConversionFromStringToSpan = false;
+
+            // When we get here we either 100% already queried for object.ToString in `ConvertConcatExprToString` (if at least 1 operand is not a string) or don't need it (if all operands are strings).
+            // Thus we can pass `isOptional` flag to avoid duplicate or unintentional "missing member" diagnostic in case we are actually missing `object.ToString()`
+            var objectToStringMethod = UnsafeGetSpecialTypeMethod(syntax, SpecialMember.System_Object__ToString, isOptional: true);
```

Comment:
>  var objectToStringMethod = UnsafeGetSpecialTypeMethod(syntax, SpecialMember.System_Object__ToString, isOptional: true); [](http://example.com/codeflow?start=10&length=121)

I think it would be better to avoid such dependency between different functions, we don't mind getting duplicate errors in such corner cases.

Reply:
Since this method symbol was only used for comparison, I can directly get it from compilation similarly as you've suggested in https://github.com/dotnet/roslyn/pull/71793#discussion_r1471717865